yapapi/engine.py

```diff
@@ -448,26 +454,109 @@ async def _process_invoices(self) -> None:
             ):
                 break
 
+    @staticmethod
+    def _reason_for_termination(
+        activity_id: str, duration: float, num_notes: int, interval: int, payable: bool
+    ):
+        freq_descr = f"{num_notes} notes/{duration}s"
+        logger.debug(
+            f"{'Payable Debit notes' if payable else 'Debit notes'} for activity {activity_id}: {freq_descr}"
+        )
+        if duration > 0 and duration + DEBIT_NOTE_INTERVAL_GRACE_PERIOD < num_notes * interval:
+            reason = {
+                "message": f"Too many {'payable ' if payable else ''}debit notes: {freq_descr} (activity: {activity_id})",
+                "golem.requestor.code": "TooManyPayableDebitNotes"
+                if payable
+                else "TooManyDebitNotes",
+            }
+            logger.error(
+                f"Too many {'payable ' if payable else ''}debit notes received."
+                " %s, activity: %s",
+                freq_descr,
+                activity_id,
+            )
+            return reason
+
+        return None
+
+    async def _verify_debit_note_interval(
+        self, agreement: Agreement, debit_note: DebitNote, duration: float
+    ):
+        interval = await agreement.get_requestor_property(PROP_DEBIT_NOTE_INTERVAL_SEC)
+        logger.debug("Debit notes interval: %ss", interval)
+        if interval:
```

Comment:
Are we sure it can't be 0?

Reply:
if the interval is 0, then it means that we're not limiting the rate so the logic still serves its purpose